sound/soc/sof/ipc.c

```diff
@@ -696,14 +697,25 @@ int snd_sof_ipc_set_get_comp_data(struct snd_sof_ipc *ipc,
 
 	/* send normal size ipc in one part */
 	if (cdata->rhdr.hdr.size <= SOF_IPC_MSG_MAX_SIZE) {
+		/* Allocate temporary buffer for new cdata */
+		reply_cdata = kzalloc(cdata->rhdr.hdr.size, GFP_KERNEL);
+		if (!reply_cdata)
+			return -ENOMEM;
+
+		/* Copy header to reply */
+		memcpy(reply_cdata, cdata, sizeof(*cdata));
+
 		err = sof_ipc_tx_message(sdev->ipc, cdata->rhdr.hdr.cmd, cdata,
-					 cdata->rhdr.hdr.size, cdata,
-					 cdata->rhdr.hdr.size);
+					 cdata->rhdr.hdr.size, reply_cdata,
+					 reply_cdata->rhdr.hdr.size);
 
 		if (err < 0)
 			dev_err(sdev->dev, "error: set/get ctrl ipc comp %d\n",
 				cdata->comp_id);
+		else if (!send)
+			memcpy(cdata, reply_cdata, reply_cdata->rhdr.hdr.size);
 
+		kfree(reply_cdata);
```

Comment:
I have absolutely zero understanding of this. I've spent 15 mn looking at the code and I guess there are assumptions I am not aware of in the IPC flow.
Someone else needs to review this, I have zero context and can't spend more time on this.

Reply:
@plbossart I have discussed the next steps for this PR with @singalsu . Basically, the direction we've taken to align all kcontrol IO to check for the status of the IPC response. The kcontrol data cache should be updated only upon a successful IPC. Seppo will work on it next week.

Reply:
@ranj063 It's not yet ready for PR but I have re-worked functions snd_sof_volume_put(), snd_sof_switch_put(), snd_sof_enum_put(), snd_sof_bytes_put(), and snd_sof_bytes_ext_put(). The IPC is first sent via a temp buffer, and copied to scontrol if return value was zero.

I tried that I still can update EQs, change volume. I need to see what else I can test with our topologies.